### ISSUE TITLE
[HOLD for payment 2024-06-03] [$250] Playback speed is not displayed correctly in video player

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#40646]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.63-7
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @blimpich / @KMichel1030 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1713565240359109

## Action Performed:
1. Upload two videos(`Video1`, `Video2`).
2. Play `Video1`.
3. Then change playback speed to `0.25` (or any other)
4. Play `Video2`.
5. Check playback speed of `Video2` (three dot menu -> playback speed).
## Expected Result: 
Playback speed of `Video2` should be displayed as `1`
## Actual Result: 
`Video2` is played at speed 1 but playback speed is displayed as `0.25`

## Workaround:
unknown
## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/Expensify/App/assets/38435837/79a8d3df-7d6a-4244-8de7-a96bb5ee10a1

https://github.com/Expensify/App/assets/38435837/81d22c16-b827-4e9b-b150-acb95d087a3e

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01648564b9cecf0115</li>
        <li>Upwork Job ID: 1781837388056326144</li>
        <li>Last Price Increase: 2024-04-24</li>
<li>Automatic offers: </li>
<ul>
<li>hungvu193 | Contributor | 0</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @stephanieelliott</details>

**Comment by KMichel1030** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Playback speed is not displayed correctly in video player

### What is the root cause of that problem?
Playback speed is stored in `currentPlaybackSpeed`. And it is not updated when playing a video.
https://github.com/Expensify/App/blob/8616678cef4af714c63abf7147853e34fb40f4ce/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx#L18

### What changes do you think we should make in order to solve the problem?
We have to get playback speed from ref and set `currentPlaybackSpeed` when we play video.
We can use `updatePlaybackSpeed` function in `VideoPopoverMenuContext` to update speed.
https://github.com/Expensify/App/blob/8616678cef4af714c63abf7147853e34fb40f4ce/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx#L22-L25

### What alternative solutions did you explore? (Optional)

**Comment by blimpich** [comment]
Reduced price as I consider this a low-priority bug.

**Comment by shahinyan11** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Playback speed is not displayed correctly in video player
### What is the root cause of that problem?
We use one [context](https://github.com/Expensify/App/blob/8616678cef4af714c63abf7147853e34fb40f4ce/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx#L15C10-L15C41)  to display playback speed for all videos. And the [currentPlaybackSpeed](https://github.com/Expensify/App/blob/8616678cef4af714c63abf7147853e34fb40f4ce/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx#L18C12-L18C32) updates there only when the select speed for some video
### What changes do you think we should make in order to solve the problem?
We should update `currentPlaybackSpeed` in context  based  current video speed  when user opens video popover.
1. Add `setCurrentPlaybackSpeed` function [here](https://github.com/Expensify/App/blob/8616678cef4af714c63abf7147853e34fb40f4ce/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx#L66C42-L66C72) 
2. Get  `setCurrentPlaybackSpeed` from  `useVideoPopoverMenuContext()` in `BaseVideoPlayer` component  and add below on this [line](https://github.com/Expensify/App/blob/8616678cef4af714c63abf7147853e34fb40f4ce/src/components/VideoPlayer/BaseVideoPlayer.tsx#L96) ( before if block )
```
currentVideoPlayerRef.current?.getStatusAsync().then((status) => {
    setCurrentPlaybackSpeed(status?.rate)
})
```

### What alternative solutions did you explore? (Optional)

**Comment by Pycomet** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Playback speed is not displayed correctly in video player - When a video's playback speed is changed, it is meant to reset when a different video is played but doesn't.

### What is the root cause of that problem?

We are using a [context manager](https://github.com/Expensify/App/blob/8616678cef4af714c63abf7147853e34fb40f4ce/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx#L18) to handle the playback speed for the videos. The issue is when the context manager handles changing the video displayed, it does not reset the video player configuration. In this case, the playback speed

### What changes do you think we should make in order to solve the problem?

The simplest solution here would be to reset the playback speed anytime the video changes. This can be archived by adding an `useEffect` hook within the context manager to do just that.

```
    useEffect(() => {
        setCurrentPlaybackSpeed(CONST.VIDEO_PLAYER.PLAYBACK_SPEEDS[2]);
    }, [currentlyPlayingURL]);
```

### What alternative solutions did you explore? (Optional)

An alternate and more robust solution would be to have a mapped state of playback speeds linked to video URLs opened on the player.

Here's what the mapper variable would look like `const [playbackSpeedMap, setPlaybackSpeedMap] = useState<{ [url: string]: PlaybackSpeed }>({});`

What this does is, it helps the application remember and attach different playback speeds to different videos. Thereby, creating a smoother user experience with the player. If a video has never been played, it has the default playback speed, but if it was changed before it will be remembered and set regardless of how many other videos you played after.

**Comment by sword1202** [comment]
Contributor details
Your Expensify account email: payoneermap1202@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~0194d962e1f9098218

**Comment by sword1202** [comment]
Contributor details
Your Expensify account email: [payoneermap1202@gmail.com](mailto:payoneermap1202@gmail.com)
Upwork Profile Link: https://www.upwork.com/freelancers/~0194d962e1f9098218
I am not getting invitation on slack channel yet.

**Comment by sword1202** [comment]
I sent a request message to your slack email, but no reponse yet.

**Comment by hungvu193** [comment]
Taking over this as C+ per this discussion

https://expensify.enterprise.slack.com/archives/C02NK2DQWUX/p1713787155736399

Cc @ntdiary

**Comment by hungvu193** [comment]
@blimpich I want to confirm the expected behavior here, if we change playback speed of `Video 1` to 0.25, then we change speed of `Video 2` to 0.5. Later when we play `Video 1`, should its playback speed be 0.25 or it will reset to 1?

**Comment by blimpich** [comment]
I would expect it to be 0.25

**Comment by hungvu193** [comment]
A friendly reminder for @shahinyan11 and @KMichel1030: please make sure you both follow the [contributing guidelines here](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#propose-a-solution-for-the-job). Otherwise, your proposal will be hidden and not reviewed.

@KMichel1030, I noticed you updated your proposal a couple of times right after @shahinyan11 posted his proposal. Your proposal also didn't elaborate on how to implement `setCurrentPlaybackSpeed`. I'll consider your proposal a duplicate of @shahinyan11's proposal.

<img width="1095" alt="Screenshot 2024-04-22 at 23 50 04" src="https://github.com/Expensify/App/assets/16502320/0eee69ac-7608-4d5d-9807-69fb22648156">

**Comment by KMichel1030** [comment]
## Proposal
[Updated] (https://github.com/Expensify/App/issues/40646#issuecomment-2067790768)

**Comment by KMichel1030** [comment]
@hungvu193
[Proposal updated](https://github.com/Expensify/App/issues/40646#issuecomment-2067790768)
Thank you for your remind and sorry for missing contributing guideline.

**Comment by shahinyan11** [comment]
@hungvu193 I apologize, but my changes are very small and just concern the writing and adding supporting links to make the proposal more understandable. And there is also no long interval between my changes

**Comment by KMichel1030** [comment]
@hungvu193 
I've reported this issue and I know about this issue well.
![image](https://github.com/Expensify/App/assets/161328186/c4eeefac-35f5-4bd8-882b-9318347b9257)

Actually, my first proposal and @shahinyan11 's proposal are exactly same.
The main point in this issue is to get playback speed from `videoPlayerRef` and set speed when opening popover menu.
`setCurrentPlaybackSpeed` is already existing function and all we have to do with `setCurrentPlaybackSpeed` is just adding it to [`contextValue`](https://github.com/KMichel1030/Expensify-App/blob/1d04d5d9560f3c6da196fdb673e8c657ffcf7744/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx#L66C11-L66C24)

**Comment by blimpich** [comment]
^^^ Ignore that, I was trying to assign @hungvu193 as the C+ for this issue since they took over. Please continue the proposal process and we can fix this later.

**Comment by hungvu193** [comment]
> @hungvu193 I apologize, but my changes are very small and just concern the writing and adding supporting links to make the proposal more understandable. And there is also no long interval between my changes

Yeah, but I still need to check every edited history, so please make sure you will leave a comment when ever you edited it 😄 .



======================================================>


> @hungvu193 I've reported this issue and I know about this issue well. ![image](https://private-user-images.githubusercontent.com/161328186/324637807-c4eeefac-35f5-4bd8-882b-9318347b9257.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM4MzU2MDQsIm5iZiI6MTcxMzgzNTMwNCwicGF0aCI6Ii8xNjEzMjgxODYvMzI0NjM3ODA3LWM0ZWVlZmFjLTM1ZjUtNGJkOC04ODJiLTkzMTgzNDdiOTI1Ny5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQyM1QwMTIxNDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01NjM2MjFiOGVlMjU5MWEzMWZkZDZhNWQ4ZTQ3YWVhNTgxMDQ5YmJiMDhjY2UyOWYwNTRiOGVlNWMzMmJmMjVlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.ztqmCWmbuJaWExrv2Wj0tLm-VbC7oR_PXu0hxoAbWAE)
> 
> Actually, my first proposal and @shahinyan11 's proposal are exactly same. The main point in this issue is to get playback speed from `videoPlayerRef` and set speed when opening popover menu. `setCurrentPlaybackSpeed` is already existing function and all we have to do with `setCurrentPlaybackSpeed` is just adding it to [`contextValue`](https://github.com/KMichel1030/Expensify-App/blob/1d04d5d9560f3c6da196fdb673e8c657ffcf7744/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx#L66C11-L66C24)

You only updated the detail of your solution after @shahinyan11 posted his solution, also I don't think your RCA is correct. Nevermind, please make sure you will leave a comment after you edited your proposal next time!

Alright guys, we have the expected [behavior here](https://github.com/Expensify/App/issues/40646#issuecomment-2070238046), please update your proposal to make sure it works!

**Comment by KMichel1030** [comment]
@hungvu193 
Thank you for your advice. I will follow contributing guide strictly from now on.

**Comment by hungvu193** [comment]
> ## Proposal
> [Updated] ([#40646 (comment)](https://github.com/Expensify/App/issues/40646#issuecomment-2067790768))

I'm sorry, I can't understand your second solution. Can you elaborate more about it? Or can you post a code diff that would be really helpful!

**Comment by KMichel1030** [comment]
@hungvu193 
Sorry, it seems I missed some test steps.
Please check this video first. I will update proposal soon.

https://github.com/Expensify/App/assets/161328186/c6a59ea9-75b7-40ea-9ea0-9cdbad62b2c3

**Comment by KMichel1030** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/40646#issuecomment-2067790768)

**Comment by hungvu193** [comment]
> ## Proposal
> [Updated](https://github.com/Expensify/App/issues/40646#issuecomment-2067790768)

Didn't work for me. Here's the diff, just to make sure we're on the same boat.

<details><summary>Diff</summary>

```diff
diff --git a/src/components/VideoPlayer/BaseVideoPlayer.tsx b/src/components/VideoPlayer/BaseVideoPlayer.tsx
index 0138e2f870..74a5697205 100644
--- a/src/components/VideoPlayer/BaseVideoPlayer.tsx
+++ b/src/components/VideoPlayer/BaseVideoPlayer.tsx
@@ -21,6 +21,7 @@ import shouldReplayVideo from './shouldReplayVideo';
 import type {VideoPlayerProps, VideoWithOnFullScreenUpdate} from './types';
 import * as VideoUtils from './utils';
 import VideoPlayerControls from './VideoPlayerControls';
+import { useVideoPopoverMenuContext } from '@components/VideoPlayerContexts/VideoPopoverMenuContext';
 
 function BaseVideoPlayer({
     url,
@@ -59,6 +60,7 @@ function BaseVideoPlayer({
         setCurrentlyPlayingURL,
     } = usePlaybackContext();
     const {isFullScreenRef} = useFullScreenContext();
+    const { playerRef } = useVideoPopoverMenuContext();
     const {isOffline} = useNetwork();
     const [duration, setDuration] = useState(videoDuration * 1000);
     const [position, setPosition] = useState(0);
@@ -93,6 +95,7 @@ function BaseVideoPlayer({
 
     const showPopoverMenu = (event?: GestureResponderEvent | KeyboardEvent) => {
         setIsPopoverVisible(true);
+        playerRef.current = videoPlayerRef.current;
         if (!event || !('nativeEvent' in event)) {
             return;
         }
diff --git a/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx b/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx
index f953ed8026..b93596dbba 100644
--- a/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx
+++ b/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx
@@ -1,4 +1,4 @@
-import React, {useCallback, useContext, useMemo, useState} from 'react';
+import React, {useCallback, useContext, useMemo, useRef, useState} from 'react';
 import * as Expensicons from '@components/Icon/Expensicons';
 import type {PopoverMenuItem} from '@components/PopoverMenu';
 import useLocalize from '@hooks/useLocalize';
@@ -7,6 +7,7 @@ import addEncryptedAuthTokenToURL from '@libs/addEncryptedAuthTokenToURL';
 import fileDownload from '@libs/fileDownload';
 import CONST from '@src/CONST';
 import type ChildrenProps from '@src/types/utils/ChildrenProps';
+import type { VideoWithOnFullScreenUpdate } from '@components/VideoPlayer/types';
 import {usePlaybackContext} from './PlaybackContext';
 import type {PlaybackSpeed, VideoPopoverMenuContext} from './types';
 
@@ -18,11 +19,12 @@ function VideoPopoverMenuContextProvider({children}: ChildrenProps) {
     const [currentPlaybackSpeed, setCurrentPlaybackSpeed] = useState<PlaybackSpeed>(CONST.VIDEO_PLAYER.PLAYBACK_SPEEDS[2]);
     const {isOffline} = useNetwork();
     const isLocalFile = currentlyPlayingURL && CONST.ATTACHMENT_LOCAL_URL_PREFIX.some((prefix) => currentlyPlayingURL.startsWith(prefix));
+    const playerRef = useRef<VideoWithOnFullScreenUpdate | null>(null);
 
     const updatePlaybackSpeed = useCallback(
         (speed: PlaybackSpeed) => {
             setCurrentPlaybackSpeed(speed);
-            currentVideoPlayerRef.current?.setStatusAsync?.({rate: speed});
+            playerRef.current?.setStatusAsync?.({rate: speed});
         },
         [currentVideoPlayerRef],
     );
@@ -63,7 +65,7 @@ function VideoPopoverMenuContextProvider({children}: ChildrenProps) {
         return items;
     }, [currentPlaybackSpeed, downloadAttachment, translate, updatePlaybackSpeed, isOffline, isLocalFile]);
 
-    const contextValue = useMemo(() => ({menuItems, updatePlaybackSpeed}), [menuItems, updatePlaybackSpeed]);
+    const contextValue = useMemo(() => ({menuItems, updatePlaybackSpeed, playerRef}), [menuItems, updatePlaybackSpeed, playerRef]);
     return <Context.Provider value={contextValue}>{children}</Context.Provider>;
 }
 

```
</details>

**Comment by KMichel1030** [comment]
@hungvu193 
I will test on my side.
Could you please add `playerRef` in dependency array of  `updatePlaybackSpeed`?

**Comment by stephanieelliott** [comment]
Reapplying the `Bug` label to get another BZ member on this while I am OOO til May 2. Thanks @bfitzexpensify,  I'll grab this back from you when I return!

**Comment by shahinyan11** [comment]
> Alright guys, we have the expected [behavior here](https://github.com/Expensify/App/issues/40646#issuecomment-2070238046), please update your proposal to make sure it works!

@hungvu193 The solution I my proposal works like this

**Comment by hungvu193** [comment]
> > Alright guys, we have the expected [behavior here](https://github.com/Expensify/App/issues/40646#issuecomment-2070238046), please update your proposal to make sure it works!
> 
> @hungvu193 The solution I my proposal works like this

I don't think it works

https://github.com/Expensify/App/assets/16502320/95f5cc12-6e1a-40ec-8388-e3fb0c5b8b81

**Comment by shahinyan11** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/40646#issuecomment-2067986160). Updated to work as [expected](https://github.com/Expensify/App/issues/40646#issuecomment-2070238046) . @hungvu193  Please check

**Comment by hungvu193** [comment]
> ## Proposal
> [Updated](https://github.com/Expensify/App/issues/40646#issuecomment-2067986160). Updated to work as [expected](https://github.com/Expensify/App/issues/40646#issuecomment-2070238046) . @hungvu193 Please check


https://github.com/Expensify/App/assets/16502320/dfaa6181-1336-4ddf-aaa3-8ab957ce3904

Please test it carefully, it still didn't work.

**Comment by shahinyan11** [comment]
@hungvu193 I tested and it works correct. Are you sure you did  all changes correct?

**Comment by shahinyan11** [comment]
> Can you confirm if this is your diff?

Yes

**Comment by KMichel1030** [comment]
@hungvu193 
Could you please let me know your opinion about my proposal(2 problems I've posted)?
It seems we have to fix both of them.

**Comment by hungvu193** [comment]
> > Can you confirm if this is your diff?
> 
> Yes


https://github.com/Expensify/App/assets/16502320/1e8ec948-2809-4274-b3e6-37ec6bc294f9

Check out this case:
- Click play video 1
- Click play video 2
- Change the play back speed of video 1
- Notice it actually changed the playback speed of video 2.

**Comment by KMichel1030** [comment]
> @hungvu193 I will test on my side. Could you please add playerRef in dependency array of updatePlaybackSpeed?
> 
> I did, but it still the same.

@hungvu193 
I've tested [similar problem ](https://github.com/Expensify/App/issues/40647#issuecomment-2067802332) before and it worked properly.
Currently, my mac is corrupted unexpectedly, so I can't test.
I'm very sorry but could you please take one more try for me?

Could you please add below part [here](https://github.com/Expensify/App/blob/64695c850dfd7dc054823b69f35c9511cd934cdd/src/components/VideoPlayerContexts/types.ts#L32)? 
`playerRef: MutableRefObject<VideoWithOnFullScreenUpdate | null>;`

**Comment by shahinyan11** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/40646#issuecomment-2067986160). Updated to cover [this](https://github.com/Expensify/App/issues/40646#issuecomment-2071834435) bug. @hungvu193 please take a look

**Comment by hungvu193** [comment]
Thanks everyone! I'm AFK for a while but will review ASAP!

**Comment by hungvu193** [comment]
Thanks @KMichel1030, your suggestion only added type to `VideoPopoverMenuContext`, it won't change the result I tested.

@shahinyan11 's [proposal here](https://github.com/Expensify/App/issues/40646#issuecomment-2067986160) looks good to me.  🎀 👀 🎀  C+ reviewed.

Given the effort everyone is putting in here, may I request a raise for this issue? @blimpich, I think $250 is a reasonable amount. Please let me know what's your thought 🙇

**Comment by KMichel1030** [comment]
@hungvu193 
I've test my proposal on my side it is working correctly.

**Comment by KMichel1030** [comment]
@hungvu193 
I've just fixed my dev env and tested.

https://github.com/Expensify/App/assets/161328186/ffbf4f03-3564-40ed-b56a-5a32eab30fb3

**Comment by hungvu193** [comment]
Hey @KMichel1030. Thanks for letting me know but I prefer using the current existing references rather than adding a new one to keep it more understandable. Anyway, let here some thoughts from @blimpich  as well!

**Comment by KMichel1030** [comment]
@hungvu193 
@shahinyan11's proposal has some problems.
Let's check this test case.
1. Upload 2 videos.
2. Play `Video1`.
3. Play `Video2`.
4. Change speed of `Video1` with out play.
5. Then try to pause `Video2`.

In above test case, `Video2` is playing but we lose control(because we control playing video with `currentVideoPlayerRef`). We can't stop `Video2`.

In my opinion, changing speed of non-playing video should not affect to playing video.
Anyway, if you prefer to change currently playing video when changing speed of non-playing video.
I've added solution as alternative solution in my proposal.

Please check my [proposal](https://github.com/Expensify/App/issues/40646#issuecomment-2067790768).

**Comment by hungvu193** [comment]
> @hungvu193 @shahinyan11's proposal has some problems. Let's check this test case.
> 
> 1. Upload 2 videos.
> 2. Play `Video1`.
> 3. Play `Video2`.
> 4. Change speed of `Video1` with out play.
> 5. Then try to pause `Video2`.
> 
> In above test case, `Video2` is playing but we lose control(because we control playing video with `currentVideoPlayerRef`). We can't stop `Video2`.
> 
> In my opinion, changing speed of non-playing video should not affect to playing video. Anyway, if you prefer to change currently playing video when changing speed of non-playing video. I've added solution as alternative solution in my proposal.
> 
> Please check my [proposal](https://github.com/Expensify/App/issues/40646#issuecomment-2067790768).

Nice catch, it seems we also need to update the `currentVideoPlayerRef` in some more specific cases.

In the mean time, I'll test your proposal again.

**Comment by KMichel1030** [comment]
@hungvu193 
> In my opinion, changing speed of non-playing video should not affect to playing video. 

Could you please let me know your suggestion about my opinion?

**Comment by hungvu193** [comment]
I'm sorry. Please test your proposal carefully, It doesn't work for me, checkout the test case below.

https://github.com/Expensify/App/assets/16502320/23dda5a8-985d-4a08-b751-d03fdeb2eeec

**Comment by KMichel1030** [comment]
@hungvu193 
Could you please show your code diff?
It's working properly on my end.

**Comment by hungvu193** [comment]
> @hungvu193 Could you please show your code diff? It's working properly on my end.

Here you are, if you find somethings wrong, please let me know.

<details><summary>Diff</summary>

```diff

diff --git a/src/components/VideoPlayer/BaseVideoPlayer.tsx b/src/components/VideoPlayer/BaseVideoPlayer.tsx
index 066dbbff3b..b58e567f62 100644
--- a/src/components/VideoPlayer/BaseVideoPlayer.tsx
+++ b/src/components/VideoPlayer/BaseVideoPlayer.tsx
@@ -17,6 +17,7 @@ import useThemeStyles from '@hooks/useThemeStyles';
 import addEncryptedAuthTokenToURL from '@libs/addEncryptedAuthTokenToURL';
 import * as DeviceCapabilities from '@libs/DeviceCapabilities';
 import CONST from '@src/CONST';
+import { useVideoPopoverMenuContext } from '@components/VideoPlayerContexts/VideoPopoverMenuContext';
 import shouldReplayVideo from './shouldReplayVideo';
 import type {VideoPlayerProps, VideoWithOnFullScreenUpdate} from './types';
 import * as VideoUtils from './utils';
@@ -59,6 +60,7 @@ function BaseVideoPlayer({
         setCurrentlyPlayingURL,
     } = usePlaybackContext();
     const {isFullScreenRef} = useFullScreenContext();
+    const {playerRef} = useVideoPopoverMenuContext();
     const {isOffline} = useNetwork();
     const [duration, setDuration] = useState(videoDuration * 1000);
     const [position, setPosition] = useState(0);
@@ -93,6 +95,8 @@ function BaseVideoPlayer({
 
     const showPopoverMenu = (event?: GestureResponderEvent | KeyboardEvent) => {
         setIsPopoverVisible(true);
+        playerRef.current = videoPlayerRef.current;
+
         if (!event || !('nativeEvent' in event)) {
             return;
         }
diff --git a/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx b/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx
index f953ed8026..3e0720bf0e 100644
--- a/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx
+++ b/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx
@@ -1,4 +1,4 @@
-import React, {useCallback, useContext, useMemo, useState} from 'react';
+import React, {useCallback, useContext, useMemo, useRef, useState} from 'react';
 import * as Expensicons from '@components/Icon/Expensicons';
 import type {PopoverMenuItem} from '@components/PopoverMenu';
 import useLocalize from '@hooks/useLocalize';
@@ -7,6 +7,7 @@ import addEncryptedAuthTokenToURL from '@libs/addEncryptedAuthTokenToURL';
 import fileDownload from '@libs/fileDownload';
 import CONST from '@src/CONST';
 import type ChildrenProps from '@src/types/utils/ChildrenProps';
+import type { VideoWithOnFullScreenUpdate } from '@components/VideoPlayer/types';
 import {usePlaybackContext} from './PlaybackContext';
 import type {PlaybackSpeed, VideoPopoverMenuContext} from './types';
 
@@ -18,13 +19,14 @@ function VideoPopoverMenuContextProvider({children}: ChildrenProps) {
     const [currentPlaybackSpeed, setCurrentPlaybackSpeed] = useState<PlaybackSpeed>(CONST.VIDEO_PLAYER.PLAYBACK_SPEEDS[2]);
     const {isOffline} = useNetwork();
     const isLocalFile = currentlyPlayingURL && CONST.ATTACHMENT_LOCAL_URL_PREFIX.some((prefix) => currentlyPlayingURL.startsWith(prefix));
+    const playerRef = useRef<VideoWithOnFullScreenUpdate | null>(null);
 
     const updatePlaybackSpeed = useCallback(
         (speed: PlaybackSpeed) => {
             setCurrentPlaybackSpeed(speed);
-            currentVideoPlayerRef.current?.setStatusAsync?.({rate: speed});
+            playerRef.current?.setStatusAsync?.({rate: speed});
         },
-        [currentVideoPlayerRef],
+        [playerRef],
     );
 
     const downloadAttachment = useCallback(() => {
@@ -63,7 +65,7 @@ function VideoPopoverMenuContextProvider({children}: ChildrenProps) {
         return items;
     }, [currentPlaybackSpeed, downloadAttachment, translate, updatePlaybackSpeed, isOffline, isLocalFile]);
 
-    const contextValue = useMemo(() => ({menuItems, updatePlaybackSpeed}), [menuItems, updatePlaybackSpeed]);
+    const contextValue = useMemo(() => ({menuItems, updatePlaybackSpeed, playerRef}), [menuItems, updatePlaybackSpeed, playerRef]);
     return <Context.Provider value={contextValue}>{children}</Context.Provider>;
 }
 
diff --git a/src/components/VideoPlayerContexts/types.ts b/src/components/VideoPlayerContexts/types.ts
index ff8d9378ca..8d4d2ac6f0 100644
--- a/src/components/VideoPlayerContexts/types.ts
+++ b/src/components/VideoPlayerContexts/types.ts
@@ -29,6 +29,7 @@ type VolumeContext = {
 type VideoPopoverMenuContext = {
     menuItems: PopoverMenuItem[];
     updatePlaybackSpeed: (speed: PlaybackSpeed) => void;
+    playerRef: MutableRefObject<VideoWithOnFullScreenUpdate | null>;
 };
 
 type FullScreenContext = {


```
</details>

**Comment by KMichel1030** [comment]
@hungvu193 
Could you please add below part [here](https://github.com/Expensify/App/blob/fd8ef990c108ed939cd6a1fd49027beb11e0570b/src/components/VideoPlayer/BaseVideoPlayer.tsx#L96)?
```
videoPlayerRef.current?.getStatusAsync().then((status) => {
            if (!('rate' in status && status.rate)) {
                return;
            }
            setCurrentPlaybackSpeed(status.rate as PlaybackSpeed);
        });
```

I'm sorry I added code diffs separately in my proposal.

**Comment by hungvu193** [comment]
Doesn't work at all!. I'm gonna postpone  reviewing your proposal for couple of hours until you make sure you test it carefully! Thanks for understanding!

https://github.com/Expensify/App/assets/16502320/15a64fb7-f161-4796-9608-d506f7234eab

**Comment by KMichel1030** [comment]
@hungvu193
Thank you for your effort.
Could you please check this video? 


https://github.com/Expensify/App/assets/161328186/c0fc96bb-28d3-4f73-a748-dfa19ba0bda0



<details>
<summary>Code Diff</summary>

```diff
diff --git a/src/components/VideoPlayer/BaseVideoPlayer.tsx b/src/components/VideoPlayer/BaseVideoPlayer.tsx
index 0138e2f870..9098ccf86c 100644
--- a/src/components/VideoPlayer/BaseVideoPlayer.tsx
+++ b/src/components/VideoPlayer/BaseVideoPlayer.tsx
@@ -10,6 +10,8 @@ import Hoverable from '@components/Hoverable';
 import PressableWithoutFeedback from '@components/Pressable/PressableWithoutFeedback';
 import {useFullScreenContext} from '@components/VideoPlayerContexts/FullScreenContext';
 import {usePlaybackContext} from '@components/VideoPlayerContexts/PlaybackContext';
+import type { PlaybackSpeed } from '@components/VideoPlayerContexts/types';
+import {useVideoPopoverMenuContext} from '@components/VideoPlayerContexts/VideoPopoverMenuContext';
 import {useVolumeContext} from '@components/VideoPlayerContexts/VolumeContext';
 import VideoPopoverMenu from '@components/VideoPopoverMenu';
 import useNetwork from '@hooks/useNetwork';
@@ -79,6 +81,7 @@ function BaseVideoPlayer({
     const isUploading = CONST.ATTACHMENT_LOCAL_URL_PREFIX.some((prefix) => url.startsWith(prefix));
     const videoStateRef = useRef<AVPlaybackStatus | null>(null);
     const {updateVolume} = useVolumeContext();
+    const {playerRef, setCurrentPlaybackSpeed} = useVideoPopoverMenuContext();
 
     const togglePlayCurrentVideo = useCallback(() => {
         videoResumeTryNumber.current = 0;
@@ -93,6 +96,13 @@ function BaseVideoPlayer({
 
     const showPopoverMenu = (event?: GestureResponderEvent | KeyboardEvent) => {
         setIsPopoverVisible(true);
+        playerRef.current = videoPlayerRef.current;
+        videoPlayerRef.current?.getStatusAsync().then((status) => {
+            if (!('rate' in status && status.rate)) {
+                return;
+            }
+            setCurrentPlaybackSpeed(status.rate as PlaybackSpeed);
+        });
         if (!event || !('nativeEvent' in event)) {
             return;
         }
diff --git a/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx b/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx
index f953ed8026..87d4ff9e6a 100644
--- a/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx
+++ b/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx
@@ -1,6 +1,7 @@
-import React, {useCallback, useContext, useMemo, useState} from 'react';
+import React, {useCallback, useContext, useMemo, useRef, useState} from 'react';
 import * as Expensicons from '@components/Icon/Expensicons';
 import type {PopoverMenuItem} from '@components/PopoverMenu';
+import type {VideoWithOnFullScreenUpdate} from '@components/VideoPlayer/types';
 import useLocalize from '@hooks/useLocalize';
 import useNetwork from '@hooks/useNetwork';
 import addEncryptedAuthTokenToURL from '@libs/addEncryptedAuthTokenToURL';
@@ -18,13 +19,14 @@ function VideoPopoverMenuContextProvider({children}: ChildrenProps) {
     const [currentPlaybackSpeed, setCurrentPlaybackSpeed] = useState<PlaybackSpeed>(CONST.VIDEO_PLAYER.PLAYBACK_SPEEDS[2]);
     const {isOffline} = useNetwork();
     const isLocalFile = currentlyPlayingURL && CONST.ATTACHMENT_LOCAL_URL_PREFIX.some((prefix) => currentlyPlayingURL.startsWith(prefix));
+    const playerRef = useRef<VideoWithOnFullScreenUpdate | null>(null);
 
     const updatePlaybackSpeed = useCallback(
         (speed: PlaybackSpeed) => {
             setCurrentPlaybackSpeed(speed);
-            currentVideoPlayerRef.current?.setStatusAsync?.({rate: speed});
+            playerRef.current?.setStatusAsync?.({rate: speed});
         },
-        [currentVideoPlayerRef],
+        [playerRef],
     );
 
     const downloadAttachment = useCallback(() => {
@@ -63,7 +65,7 @@ function VideoPopoverMenuContextProvider({children}: ChildrenProps) {
         return items;
     }, [currentPlaybackSpeed, downloadAttachment, translate, updatePlaybackSpeed, isOffline, isLocalFile]);
 
-    const contextValue = useMemo(() => ({menuItems, updatePlaybackSpeed}), [menuItems, updatePlaybackSpeed]);
+    const contextValue = useMemo(() => ({menuItems, playerRef, updatePlaybackSpeed, setCurrentPlaybackSpeed}), [menuItems, playerRef, updatePlaybackSpeed, setCurrentPlaybackSpeed]);
     return <Context.Provider value={contextValue}>{children}</Context.Provider>;
 }
 
diff --git a/src/components/VideoPlayerContexts/types.ts b/src/components/VideoPlayerContexts/types.ts
index ff8d9378ca..9dc5cbcfd6 100644
--- a/src/components/VideoPlayerContexts/types.ts
+++ b/src/components/VideoPlayerContexts/types.ts
@@ -28,7 +28,9 @@ type VolumeContext = {
 
 type VideoPopoverMenuContext = {
     menuItems: PopoverMenuItem[];
+    playerRef: MutableRefObject<VideoWithOnFullScreenUpdate | null>;
     updatePlaybackSpeed: (speed: PlaybackSpeed) => void;
+    setCurrentPlaybackSpeed: (speed: PlaybackSpeed) => void;
 };
 
 type FullScreenContext = {
```

</details>

**Comment by shahinyan11** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/40646#issuecomment-2067986160). Updated to cover [this](https://github.com/Expensify/App/issues/40646#issuecomment-2073933839) bug. @hungvu193 please take a look

**Comment by shahinyan11** [comment]
> Given the effort everyone is putting in here, may I request a raise for this issue? @blimpich, I think $250 is a reasonable amount. Please let me know what's your thought 🙇

I agree with this too

**Comment by hungvu193** [comment]
> ## Proposal
> [Updated](https://github.com/Expensify/App/issues/40646#issuecomment-2067986160). Added set `3` to cover [this](https://github.com/Expensify/App/issues/40646#issuecomment-2073933839) bug. @hungvu193 please take a look

@shahinyan11  Please checkout the failed case below:

<details><summary>Result</summary>

https://github.com/Expensify/App/assets/16502320/5a734692-b36d-41b2-9c38-2a77d257cb5e

</details>

<br/>

<br />

>Thank you for your effort.
> Could you please check this video and code diff?


Thanks for your explanation @KMichel1030. Please write your proposal more cleanly next time. Your proposal worked; however, I don't like the idea of creating a new playerRef, as that will make the code hard to understand. The ideal solution should revolve around our current existing video ref.

**Comment by KMichel1030** [comment]
@hungvu193 

> Please write your proposal more cleanly next time.

Thank your for your advice and check.

--------------------------------------------------------------------------

> @shahinyan11 Please checkout the failed case below:

It seems we can't resolve this problem without creating new ref.

If you prefer using existing refs, we have to pause previous playing video and set `currentVideoPlayerRef` when opening popover menu of non-playing video.

I've added solution as alternative solution in my proposal([Updated](https://github.com/Expensify/App/issues/40646#issuecomment-2067790768)).

Could you please check?

**Comment by hungvu193** [comment]
> I've added solution as alternative solution in my proposal([Updated](https://github.com/Expensify/App/issues/40646#issuecomment-2067790768)).
> 
> Could you please check?

I'm sorry, may be I have bad english, can you post your code diff of your alt solution here? Thank you

**Comment by KMichel1030** [comment]
@hungvu193 
This is code diff of my alt solution. Please check.

<details>
<summary>Code Diff</summary>

```diff
diff --git a/src/components/VideoPlayer/BaseVideoPlayer.tsx b/src/components/VideoPlayer/BaseVideoPlayer.tsx
index 9380ce43c4..8f67fae551 100644
--- a/src/components/VideoPlayer/BaseVideoPlayer.tsx
+++ b/src/components/VideoPlayer/BaseVideoPlayer.tsx
@@ -10,6 +10,7 @@ import Hoverable from '@components/Hoverable';
 import PressableWithoutFeedback from '@components/Pressable/PressableWithoutFeedback';
 import {useFullScreenContext} from '@components/VideoPlayerContexts/FullScreenContext';
 import {usePlaybackContext} from '@components/VideoPlayerContexts/PlaybackContext';
+import {useVideoPopoverMenuContext} from '@components/VideoPlayerContexts/VideoPopoverMenuContext';
 import VideoPopoverMenu from '@components/VideoPopoverMenu';
 import useNetwork from '@hooks/useNetwork';
 import useThemeStyles from '@hooks/useThemeStyles';
@@ -67,6 +68,7 @@ function BaseVideoPlayer({
     const isCurrentlyURLSet = currentlyPlayingURL === url;
     const isUploading = CONST.ATTACHMENT_LOCAL_URL_PREFIX.some((prefix) => url.startsWith(prefix));
     const videoStateRef = useRef<AVPlaybackStatus | null>(null);
+    const {setCurrentPlaybackSpeed} = useVideoPopoverMenuContext();
 
     const togglePlayCurrentVideo = useCallback(() => {
         videoResumeTryNumber.current = 0;
@@ -81,6 +83,10 @@ function BaseVideoPlayer({
 
     const showPopoverMenu = (event?: GestureResponderEvent | KeyboardEvent) => {
         setIsPopoverVisible(true);
+        updateCurrentlyPlayingURL(url);
+        videoPlayerRef.current?.getStatusAsync().then((status) => {
+            setCurrentPlaybackSpeed(status?.rate)
+        })
         if (!event || !('nativeEvent' in event)) {
             return;
         }
@@ -186,7 +192,7 @@ function BaseVideoPlayer({
         if (shouldUseSharedVideoElement || url !== currentlyPlayingURL || isFullScreenRef.current) {
             return;
         }
-        shareVideoPlayerElements(videoPlayerRef.current, videoPlayerElementParentRef.current, videoPlayerElementRef.current, isUploading);
+        shareVideoPlayerElements(videoPlayerRef.current, videoPlayerElementParentRef.current, videoPlayerElementRef.current, !isUploading && !isPopoverVisible);
     }, [currentlyPlayingURL, shouldUseSharedVideoElement, shareVideoPlayerElements, url, isUploading, isFullScreenRef]);
 
     // append shared video element to new parent (used for example in attachment modal)
diff --git a/src/components/VideoPlayerContexts/PlaybackContext.tsx b/src/components/VideoPlayerContexts/PlaybackContext.tsx
index 5306dccb56..a5d1dcf525 100644
--- a/src/components/VideoPlayerContexts/PlaybackContext.tsx
+++ b/src/components/VideoPlayerContexts/PlaybackContext.tsx
@@ -49,12 +49,12 @@ function PlaybackContextProvider({children}: ChildrenProps) {
     );
 
     const shareVideoPlayerElements = useCallback(
-        (ref: VideoWithOnFullScreenUpdate | null, parent: View | HTMLDivElement | null, child: View | HTMLDivElement | null, isUploading: boolean) => {
+        (ref: VideoWithOnFullScreenUpdate | null, parent: View | HTMLDivElement | null, child: View | HTMLDivElement | null, shouldPlay: boolean) => {
             currentVideoPlayerRef.current = ref;
             setOriginalParent(parent);
             setSharedElement(child);
             // Prevents autoplay when uploading the attachment
-            if (!isUploading) {
+            if (shouldPlay) {
                 playVideo();
             }
         },
diff --git a/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx b/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx
index f953ed8026..5b3c81dc7d 100644
--- a/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx
+++ b/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx
@@ -63,7 +63,7 @@ function VideoPopoverMenuContextProvider({children}: ChildrenProps) {
         return items;
     }, [currentPlaybackSpeed, downloadAttachment, translate, updatePlaybackSpeed, isOffline, isLocalFile]);
 
-    const contextValue = useMemo(() => ({menuItems, updatePlaybackSpeed}), [menuItems, updatePlaybackSpeed]);
+    const contextValue = useMemo(() => ({menuItems, updatePlaybackSpeed, setCurrentPlaybackSpeed}), [menuItems, updatePlaybackSpeed, setCurrentPlaybackSpeed]);
     return <Context.Provider value={contextValue}>{children}</Context.Provider>;
 }
 
diff --git a/src/components/VideoPlayerContexts/types.ts b/src/components/VideoPlayerContexts/types.ts
index 3dd2979519..680459bd9f 100644
--- a/src/components/VideoPlayerContexts/types.ts
+++ b/src/components/VideoPlayerContexts/types.ts
@@ -27,6 +27,7 @@ type VolumeContext = {
 type VideoPopoverMenuContext = {
     menuItems: PopoverMenuItem[];
     updatePlaybackSpeed: (speed: PlaybackSpeed) => void;
+    setCurrentPlaybackSpeed: (speed: PlaybackSpeed) => void;
 };
 
 type FullScreenContext = {
```

</details>

**Comment by hungvu193** [comment]
> @hungvu193 This is code diff of my alt solution. Please check.
> 
> Code Diff

Failed case:

https://github.com/Expensify/App/assets/16502320/bdba1387-1a37-403a-a107-419ea7889a48

**Comment by hungvu193** [comment]
Bump on [slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1713957511078099) for more proposals.

**Comment by KMichel1030** [comment]
@hungvu193 
Could you please show your code diff?
It's working properly on my side.

**Comment by hungvu193** [comment]
> @hungvu193
> Could you please show your code diff?

<details><summary>Diff</summary>

```diff
diff --git a/src/components/VideoPlayer/BaseVideoPlayer.tsx b/src/components/VideoPlayer/BaseVideoPlayer.tsx
index 066dbbff3b..80b3648c4b 100644
--- a/src/components/VideoPlayer/BaseVideoPlayer.tsx
+++ b/src/components/VideoPlayer/BaseVideoPlayer.tsx
@@ -17,11 +17,13 @@ import useThemeStyles from '@hooks/useThemeStyles';
 import addEncryptedAuthTokenToURL from '@libs/addEncryptedAuthTokenToURL';
 import * as DeviceCapabilities from '@libs/DeviceCapabilities';
 import CONST from '@src/CONST';
+import { useVideoPopoverMenuContext } from '@components/VideoPlayerContexts/VideoPopoverMenuContext';
 import shouldReplayVideo from './shouldReplayVideo';
 import type {VideoPlayerProps, VideoWithOnFullScreenUpdate} from './types';
 import * as VideoUtils from './utils';
 import VideoPlayerControls from './VideoPlayerControls';
 
+
 function BaseVideoPlayer({
     url,
     resizeMode = ResizeMode.CONTAIN,
@@ -58,6 +60,7 @@ function BaseVideoPlayer({
         videoResumeTryNumber,
         setCurrentlyPlayingURL,
     } = usePlaybackContext();
+    const {setCurrentPlaybackSpeed} = useVideoPopoverMenuContext();
     const {isFullScreenRef} = useFullScreenContext();
     const {isOffline} = useNetwork();
     const [duration, setDuration] = useState(videoDuration * 1000);
@@ -93,6 +96,10 @@ function BaseVideoPlayer({
 
     const showPopoverMenu = (event?: GestureResponderEvent | KeyboardEvent) => {
         setIsPopoverVisible(true);
+           updateCurrentlyPlayingURL(url);
+        videoPlayerRef.current?.getStatusAsync().then((status) => {
+            setCurrentPlaybackSpeed(status?.rate)
+        })
         if (!event || !('nativeEvent' in event)) {
             return;
         }
@@ -233,8 +240,8 @@ function BaseVideoPlayer({
         if (shouldUseSharedVideoElement || url !== currentlyPlayingURL || isFullScreenRef.current) {
             return;
         }
-        shareVideoPlayerElements(videoPlayerRef.current, videoPlayerElementParentRef.current, videoPlayerElementRef.current, isUploading);
-    }, [currentlyPlayingURL, shouldUseSharedVideoElement, shareVideoPlayerElements, url, isUploading, isFullScreenRef]);
+        shareVideoPlayerElements(videoPlayerRef.current, videoPlayerElementParentRef.current, videoPlayerElementRef.current, !isUploading && !isPopoverVisible);
+    }, [currentlyPlayingURL, shouldUseSharedVideoElement, shareVideoPlayerElements, url, isUploading, isFullScreenRef, isPopoverVisible]);
 
     // append shared video element to new parent (used for example in attachment modal)
     useEffect(() => {
diff --git a/src/components/VideoPlayerContexts/PlaybackContext.tsx b/src/components/VideoPlayerContexts/PlaybackContext.tsx
index 499dd2b07f..c279739d5e 100644
--- a/src/components/VideoPlayerContexts/PlaybackContext.tsx
+++ b/src/components/VideoPlayerContexts/PlaybackContext.tsx
@@ -51,12 +51,12 @@ function PlaybackContextProvider({children}: ChildrenProps) {
     );
 
     const shareVideoPlayerElements = useCallback(
-        (ref: VideoWithOnFullScreenUpdate | null, parent: View | HTMLDivElement | null, child: View | HTMLDivElement | null, isUploading: boolean) => {
+        (ref: VideoWithOnFullScreenUpdate | null, parent: View | HTMLDivElement | null, child: View | HTMLDivElement | null, shouldPlay: boolean) => {
             currentVideoPlayerRef.current = ref;
             setOriginalParent(parent);
             setSharedElement(child);
             // Prevents autoplay when uploading the attachment
-            if (!isUploading) {
+            if (shouldPlay) {
                 playVideo();
             }
         },
diff --git a/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx b/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx
index f953ed8026..5b3c81dc7d 100644
--- a/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx
+++ b/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx
@@ -63,7 +63,7 @@ function VideoPopoverMenuContextProvider({children}: ChildrenProps) {
         return items;
     }, [currentPlaybackSpeed, downloadAttachment, translate, updatePlaybackSpeed, isOffline, isLocalFile]);
 
-    const contextValue = useMemo(() => ({menuItems, updatePlaybackSpeed}), [menuItems, updatePlaybackSpeed]);
+    const contextValue = useMemo(() => ({menuItems, updatePlaybackSpeed, setCurrentPlaybackSpeed}), [menuItems, updatePlaybackSpeed, setCurrentPlaybackSpeed]);
     return <Context.Provider value={contextValue}>{children}</Context.Provider>;
 }
 


```</details>

**Comment by shahinyan11** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/40646#issuecomment-2067986160). Update set `3` to cover [this](https://github.com/Expensify/App/issues/40646#issuecomment-2074581027) bug. @hungvu193 please take a look

**Comment by shahinyan11** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/40646#issuecomment-2067986160). Update to cover all bugs mentioned above . @hungvu193 please take a look

**Comment by hungvu193** [comment]
> @hungvu193 Could you please remove `isPopoverVisible` in below part?
> 
> ![image](https://private-user-images.githubusercontent.com/161328186/325211925-110d0f31-2252-483d-b966-d1976d9b8545.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQwMDk2MzcsIm5iZiI6MTcxNDAwOTMzNywicGF0aCI6Ii8xNjEzMjgxODYvMzI1MjExOTI1LTExMGQwZjMxLTIyNTItNDgzZC1iOTY2LWQxOTc2ZDliODU0NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQyNVQwMTQyMTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iZGQyODFkMDE1OWNmZWNkYWZhOWVjNjNiZDhiNTZmOGQzZGE1MDkzNTdlYTZjMmIwOWRiYTU0YzdiZjc2YjVlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.J5Z-PvxRCA3iGZ2qe0dBYEa43q_2eKHpbCODddHM1EM)
> 
> ```diff
> +    }, [currentlyPlayingURL, shouldUseSharedVideoElement, shareVideoPlayerElements, url, isUploading, isFullScreenRef, isPopoverVisible]);
> ```

I'm sorry @KMichel1030 I still got the same result.

@shahinyan11 Your updated proposal is now a dupe of @KMichel1030 proposal.

**Comment by KMichel1030** [comment]
@hungvu193 
I will share my code diff soon.

**Comment by hungvu193** [comment]
> @hungvu193 I will share my code diff soon.

And don't forget to update your proposal along with it. Thank you/

**Comment by KMichel1030** [comment]
@hungvu193 
Could you please check this code diff? I've checked on my side.

<details>
<summary>Code Diff</summary>

```diff

diff --git a/src/components/VideoPlayer/BaseVideoPlayer.tsx b/src/components/VideoPlayer/BaseVideoPlayer.tsx
index 9380ce43c4..a8db4b827b 100644
--- a/src/components/VideoPlayer/BaseVideoPlayer.tsx
+++ b/src/components/VideoPlayer/BaseVideoPlayer.tsx
@@ -10,6 +10,7 @@ import Hoverable from '@components/Hoverable';
 import PressableWithoutFeedback from '@components/Pressable/PressableWithoutFeedback';
 import {useFullScreenContext} from '@components/VideoPlayerContexts/FullScreenContext';
 import {usePlaybackContext} from '@components/VideoPlayerContexts/PlaybackContext';
+import {useVideoPopoverMenuContext} from '@components/VideoPlayerContexts/VideoPopoverMenuContext';
 import VideoPopoverMenu from '@components/VideoPopoverMenu';
 import useNetwork from '@hooks/useNetwork';
 import useThemeStyles from '@hooks/useThemeStyles';
@@ -67,6 +68,7 @@ function BaseVideoPlayer({
     const isCurrentlyURLSet = currentlyPlayingURL === url;
     const isUploading = CONST.ATTACHMENT_LOCAL_URL_PREFIX.some((prefix) => url.startsWith(prefix));
     const videoStateRef = useRef<AVPlaybackStatus | null>(null);
+    const {setCurrentPlaybackSpeed} = useVideoPopoverMenuContext();
 
     const togglePlayCurrentVideo = useCallback(() => {
         videoResumeTryNumber.current = 0;
@@ -81,6 +83,10 @@ function BaseVideoPlayer({
 
     const showPopoverMenu = (event?: GestureResponderEvent | KeyboardEvent) => {
         setIsPopoverVisible(true);
+        updateCurrentlyPlayingURL(url);
+        videoPlayerRef.current?.getStatusAsync().then((status) => {
+            setCurrentPlaybackSpeed(status?.rate)
+        })
         if (!event || !('nativeEvent' in event)) {
             return;
         }
@@ -182,11 +188,12 @@ function BaseVideoPlayer({
     }, [url, currentVideoPlayerRef, isUploading]);
 
     // update shared video elements
+    //eslint-disable-next-line react-hooks/exhaustive-deps
     useEffect(() => {
         if (shouldUseSharedVideoElement || url !== currentlyPlayingURL || isFullScreenRef.current) {
             return;
         }
-        shareVideoPlayerElements(videoPlayerRef.current, videoPlayerElementParentRef.current, videoPlayerElementRef.current, isUploading);
+        shareVideoPlayerElements(videoPlayerRef.current, videoPlayerElementParentRef.current, videoPlayerElementRef.current, !isUploading && !isPopoverVisible);
     }, [currentlyPlayingURL, shouldUseSharedVideoElement, shareVideoPlayerElements, url, isUploading, isFullScreenRef]);
 
     // append shared video element to new parent (used for example in attachment modal)
diff --git a/src/components/VideoPlayerContexts/PlaybackContext.tsx b/src/components/VideoPlayerContexts/PlaybackContext.tsx
index 5306dccb56..a5d1dcf525 100644
--- a/src/components/VideoPlayerContexts/PlaybackContext.tsx
+++ b/src/components/VideoPlayerContexts/PlaybackContext.tsx
@@ -49,12 +49,12 @@ function PlaybackContextProvider({children}: ChildrenProps) {
     );
 
     const shareVideoPlayerElements = useCallback(
-        (ref: VideoWithOnFullScreenUpdate | null, parent: View | HTMLDivElement | null, child: View | HTMLDivElement | null, isUploading: boolean) => {
+        (ref: VideoWithOnFullScreenUpdate | null, parent: View | HTMLDivElement | null, child: View | HTMLDivElement | null, shouldPlay: boolean) => {
             currentVideoPlayerRef.current = ref;
             setOriginalParent(parent);
             setSharedElement(child);
             // Prevents autoplay when uploading the attachment
-            if (!isUploading) {
+            if (shouldPlay) {
                 playVideo();
             }
         },
diff --git a/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx b/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx
index f953ed8026..5b3c81dc7d 100644
--- a/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx
+++ b/src/components/VideoPlayerContexts/VideoPopoverMenuContext.tsx
@@ -63,7 +63,7 @@ function VideoPopoverMenuContextProvider({children}: ChildrenProps) {
         return items;
     }, [currentPlaybackSpeed, downloadAttachment, translate, updatePlaybackSpeed, isOffline, isLocalFile]);
 
-    const contextValue = useMemo(() => ({menuItems, updatePlaybackSpeed}), [menuItems, updatePlaybackSpeed]);
+    const contextValue = useMemo(() => ({menuItems, updatePlaybackSpeed, setCurrentPlaybackSpeed}), [menuItems, updatePlaybackSpeed, setCurrentPlaybackSpeed]);
     return <Context.Provider value={contextValue}>{children}</Context.Provider>;
 }
 
diff --git a/src/components/VideoPlayerContexts/types.ts b/src/components/VideoPlayerContexts/types.ts
index 3dd2979519..680459bd9f 100644
--- a/src/components/VideoPlayerContexts/types.ts
+++ b/src/components/VideoPlayerContexts/types.ts
@@ -27,6 +27,7 @@ type VolumeContext = {
 type VideoPopoverMenuContext = {
     menuItems: PopoverMenuItem[];
     updatePlaybackSpeed: (speed: PlaybackSpeed) => void;
+    setCurrentPlaybackSpeed: (speed: PlaybackSpeed) => void;
 };
 
 type FullScreenContext = {

```
</details>

**Comment by hungvu193** [comment]
> @hungvu193 Could you please check this code diff? I've checked on my side.
> 
> Code Diff

Sure. CAn you please create a branch on your fork as well? just to make sure I won't miss any thing.

**Comment by KMichel1030** [comment]
@hungvu193
Could you please check [here](https://github.com/KMichel1030/Expensify-App/tree/temp)?

**Comment by hungvu193** [comment]
> @hungvu193 Could you please check [here](https://github.com/KMichel1030/Expensify-App/tree/temp)?

Sure thing. Please merge main

**Comment by KMichel1030** [comment]
@hungvu193 
Oh, I'm sorry. I found failed cases. I will check more and let you know.

**Comment by shahinyan11** [comment]
> @shahinyan11 Your updated proposal is now a dupe of @KMichel1030 's [proposal](https://github.com/Expensify/App/issues/40646#issuecomment-2067790768).

If it is dupe How does my solution work and @KMichel1030 's doesn’t? .

**Comment by hungvu193** [comment]
> Thanks for your explanation @KMichel1030. Please write your proposal more cleanly next time. Your proposal worked; however, I don't like the idea of creating a new playerRef, as that will make the code hard to understand. The ideal solution should revolve around our current existing video ref.

His solution worked. I suggested change because I prefer using existing ref rather than creating a new one.

**Comment by shahinyan11** [comment]
@hungvu193 Ok. you decide. Thx for reviewing

**Comment by shahinyan11** [comment]
@hungvu193 Could you please clarify the behavior .If the `video 2` is playing and we press on three dot of of `video 1` should `video 2` be paused ? and should `video 1` be played ?

**Comment by hungvu193** [comment]
> @hungvu193 Could you please clarify the behavior .If the `video 2` is playing and we press on three dot of `video 1` should `video 2` be paused ? and should `video 1` be played ?

Video2 should continue playing, you can check this behavior on staging. We only fix the playback speed here, other behaviors should remain the same with staging.

**Comment by KMichel1030** [comment]
@hungvu193 

> Video2 should continue playing, you can check this behavior on staging. We only fix the playback speed here, other behaviors should remain the same with staging.

My alternative solution changes currently playing video and so it's not the solution we want.
If we want fix only the playback speed and other behaviors should remain the same with staging, my main proposal is solution for that.

Let's check this test case. I think we can't resolve this problem without new ref.
1. upload 2 videos.
2. play `Video2`
3. play `Video1`
4. open three dot of `Video2` and change playback speed
5. play `Video2`

In staging server, `Video1` is paused.
- Workflow in staging
playing video's ref is saved in `currentVideoPlayerRef` so we pause previously playing video before updating `currentVideoPlayerRef` when play another video.

If we want use existing ref, we have to set `currentVideoPlayerRef` with `videoPlayerRef` (ref of video player which we try to change speed) and change speed of `currentVideoPlayerRef`. In this case, previously playing video's ref is lost so we can't pause video as it is in the staging.

Please let me know your suggestion. Thank you.

**Comment by shahinyan11** [comment]
> Video2 should continue playing, you can check this behavior on staging. We only fix the playback speed here, other behaviors should remain the same with staging.

To achieve this, the video player context must be separated from the video popover context, otherwise we may have many other problems.

**Comment by hungvu193** [comment]
Thank you both. I see the issue now. I agree that we should have a separate ref, because of that @KMichel1030 's [ main proposal](https://github.com/Expensify/App/issues/40646#issuecomment-2067790768) looks good to me!.
cc @blimpich

**Comment by blimpich** [comment]
@KMichel1030 please feel free to raise a PR. The bug zero team members will make sure you get paid once the PR is up and merged.

Please follow the steps [here](https://github.com/Expensify/App/issues/40646#issuecomment-2067790779) so that the process of assigning a contract to you can be automatic in the future.

**Comment by stephanieelliott** [comment]
I'm back from OOO, thanks for watching this while I was out @bfitzexpensify! 

This PR is under active review!

**Comment by stephanieelliott** [comment]
All conflicts are resolved, the PR is just waiting on an approving review.

**Comment by stephanieelliott** [comment]
Seems like we hit a snag with testing on some platforms, see the PR for discussion on that

**Comment by stephanieelliott** [comment]
PR merged to main, should hit staging next deploy

**Comment by KMichel1030** [comment]
Hi, @stephanieelliott 
could you please let me know if can I get paid today?

**Comment by stephanieelliott** [comment]
Hey @hungvu193 can you please complete the BZ checklist? 

@KMichel1030 yes, you can be paid today! I extended an offer to you on Upwork, can you please let me know once you've accepted? Once you do I will issue payment ASAP 
https://www.upwork.com/nx/wm/offer/102582520

**Comment by KMichel1030** [comment]
@stephanieelliott 
Thank you for offer. I've just accepted offer.

**Comment by stephanieelliott** [comment]
Summarizing payment on this issue: 

 - Contributor: @KMichel1030 $250 via Upwork  -- PAID 
 - Contributor+: @hungvu193 $250 via Upwork  -- PAID 
 
 Upwork job is here: https://www.upwork.com/jobs/~1737876297360310272

**Comment by hungvu193** [comment]
- [x] The PR that introduced the bug has been identified. Link to the PR: https://github.com/Expensify/App/pull/30829
- [x] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: https://github.com/Expensify/App/pull/30829/files#r1625244663
- [x] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: N/A
- [x] Determine if we should create a regression test for this bug: Yes

**Regression test**:
1. Upload two videos(Video1, Video2).
2. Play Video1.
3. Then change playback speed to 0.25 (or any other).
4. Play Video2.
5. Verify that playback speed of Video2 (three dot menu -> playback speed) displays correctly.

Do we 👍 or 👎 ?

**Comment by hungvu193** [comment]
We're good to close @stephanieelliott

**Comment by stephanieelliott** [comment]
Thanks @hungvu193! PR for regression test here: https://github.com/Expensify/Expensify/issues/402935
